shared/src/price_estimate.rs

```diff
@@ -205,178 +220,146 @@ impl BaselinePriceEstimator {
+            .best_execution(
+                sell_token,
+                buy_token,
+                sell_amount,
+                |amount, path, pools| {
+                    estimate_buy_amount(amount, path, pools)
+                        .map(&path_comparison)
+                        .unwrap_or_else(|| -U256::max_value().to_big_rational())
+                },
+                |amount, path, pools| {
+                    estimate_buy_amount(amount, path, pools).map(|estimate| estimate.value)
+                },
+            )
+            .await?;
+        let price = amounts_to_price(sell_amount, buy_amount)?;
+        Ok((path, price))
     }
 
-    #[async_recursion]
+    /// Returns path and price (sell_amount / buy_amount).
+    /// If sell_token_price_in_native_token is set then it will be used to take gas cost into
+    /// account.
     async fn best_execution_buy_order(
         &self,
         sell_token: H160,
         buy_token: H160,
         buy_amount: U256,
         gas_price: f64,
-        consider_gas_costs: bool,
-    ) -> Result<(Vec<H160>, U256)> {
-        let path_comparison = if consider_gas_costs {
-            // Do not consider gas costs below to avoid infinite recursion.
-            let sell_token_price_in_native_token = self
-                .estimate_price_helper(
-                    self.native_token,
-                    sell_token,
-                    self.amount_to_estimate_prices_with,
-                    OrderKind::Sell,
-                    false,
-                )
-                .await?
-                .1;
-            Box::new(move |sell_estimate: Estimate<U256, Pool>| {
-                self.evaluate_path_of_buy_order_considering_gas_costs(
-                    sell_token_price_in_native_token.clone(),
-                    gas_price,
-                    sell_estimate,
-                )
-            }) as Box<dyn Fn(Estimate<U256, Pool>) -> BigRational + Send + Sync>
-        } else {
-            Box::new(|sell_estimate: Estimate<U256, Pool>| {
-                self.evaluate_path_of_buy_order_disregarding_gas_costs(sell_estimate)
-            }) as Box<dyn Fn(Estimate<U256, Pool>) -> BigRational + Send + Sync>
+        sell_token_price_in_native_token: Option<BigRational>,
+    ) -> Result<(Vec<H160>, BigRational), PriceEstimationError> {
+        let path_comparison = |sell_estimate: Estimate<U256, Pool>| {
+            if let Some(sell_token_price_in_native_token) = &sell_token_price_in_native_token {
+                let sell_amount_in_native_token =
+                    sell_estimate.value.to_big_rational() * sell_token_price_in_native_token;
+                let tx_cost_in_native_token = U256::from_f64_lossy(gas_price).to_big_rational()
+                    * BigRational::from_integer(sell_estimate.gas_cost().into());
+                -sell_amount_in_native_token - tx_cost_in_native_token
+            } else {
+                -sell_estimate.value.to_big_rational()
+            }
```

Comment:
No big deal, but notice that here we'll have a condition repeatedly evaluated inside the function, where before the condition was factored out.

Reply:
Yes. It is avoidable but makes the code a bit uglier so I did it this way.